File: mobile/lib/core/plugin/registry.dart
import 'dart:async';

import '../transport/handshake.dart';
import '../transport/session.dart';
import 'plugin.dart';

/// What the registry needs from whatever embeds it (today, AppService) to
/// actually reach peers, persist settings, and surface events. The registry
/// is otherwise self-contained and knows nothing about Flutter widgets or
/// storage formats.
abstract class PluginHost {
  Future<void> send(String deviceId, Map<String, dynamic> message);
  void broadcast(String capability, Map<String, dynamic> message);
  List<PeerRef> connectedPeers(String capability);

  /// Whether deviceId has been granted capability — the per-device trust
  /// permission a user sets per paired device.
  bool allows(String deviceId, String capability);

  void emit(PluginEvent event);
  Future<HandshakeResult> dialTransfer(String deviceId);
  Map<String, dynamic> loadPluginSettings(PluginId id);
  Future<void> savePluginSettings(PluginId id, Map<String, dynamic> data);
}

class _Registered {
  final WeDropPlugin plugin;
  bool enabled;
  _Registered(this.plugin, this.enabled);
}

/// Holds every compiled-in plugin, routes inbound messages to the plugin
/// that claimed each message type, and lets a plugin be enabled or disabled
/// at runtime without touching any other plugin.
///
/// This is the "loadable/unloadable" mechanism: plugins are compiled into
/// the single Flutter app (Dart has no stable hot-load mechanism), but the
/// registry's enable/disable is a genuine runtime toggle — a disabled
/// plugin's messages are silently dropped and its background work is
/// stopped, without restarting the app or any other plugin.
class PluginRegistry {
  final PluginHost host;

  final Map<PluginId, _Registered> _plugins = {};
  final Map<String, PluginId> _byMsg = {};

  PluginRegistry(this.host);

  /// Adds a plugin, enabled by default unless [enabledByDefault] is false.
  /// Throws if another plugin already claimed one of its message types.
  void register(WeDropPlugin plugin, {bool enabledByDefault = true}) {
    if (_plugins.containsKey(plugin.id)) {
      throw StateError('plugin "${plugin.id}" already registered');
    }
    for (final type in plugin.messageTypes) {
      final owner = _byMsg[type];
      if (owner != null) {
        throw StateError('message type "$type" already claimed by plugin "$owner"');
      }
    }

    plugin.init(_PluginApiImpl(plugin.id, this));

    for (final type in plugin.messageTypes) {
      _byMsg[type] = plugin.id;
    }
    _plugins[plugin.id] = _Registered(plugin, enabledByDefault);
  }

  WeDropPlugin? plugin(PluginId id) => _plugins[id]?.plugin;

  List<PluginId> get ids => _plugins.keys.toList();

  bool isEnabled(PluginId id) => _plugins[id]?.enabled ?? false;

  /// Toggles a plugin at runtime. Disabling stops its background work and
  /// starts silently dropping its messages; re-enabling restarts it.
  Future<void> setEnabled(PluginId id, bool enabled) async {
    final entry = _plugins[id];
    if (entry == null) throw StateError('unknown plugin "$id"');
    if (entry.enabled == enabled) return;
    entry.enabled = enabled;
    if (enabled) {
      await entry.plugin.start();
    } else {
      entry.plugin.stop();
    }
  }

  Future<void> startAll() async {
    for (final entry in _plugins.values) {
      if (entry.enabled) await entry.plugin.start();
    }
  }

  void stopAll() {
    for (final entry in _plugins.values) {
      entry.plugin.stop();
    }
  }

  // ------------------------------------------------ SessionHandler surface

  /// Looks up which plugin claimed [msgType] and, if enabled, hands it the
  /// raw payload; an unclaimed or disabled plugin's message is dropped
  /// silently.
  void onMessage(Session session, String msgType, Map<String, dynamic> raw) {
    final id = _byMsg[msgType];
    if (id == null) return;
    final entry = _plugins[id];
    if (entry == null || !entry.enabled) return;
    unawaited(
      entry.plugin.handleMessage(
        PeerRef(session.deviceId, session.peerInfo),
        msgType,
        raw,
      ),
    );
  }

  void onPeerConnected(PeerRef peer) {
    for (final entry in _plugins.values) {
      if (entry.enabled) entry.plugin.onPeerConnected(peer);
    }
  }

  void onPeerDisconnected(String deviceId) {
    for (final entry in _plugins.values) {
      if (entry.enabled) entry.plugin.onPeerDisconnected(deviceId);
    }
  }
}

class _PluginApiImpl implements PluginApi {
  final PluginId _id;
  final PluginRegistry _registry;
  _PluginApiImpl(this._id, this._registry);

  @override
  Future<void> send(String deviceId, Map<String, dynamic> message) =>
      _registry.host.send(deviceId, message);

  @override
  void broadcast(Map<String, dynamic> message) => _registry.host.broadcast(_id, message);

  @override
  List<PeerRef> connectedPeers() => _registry.host.connectedPeers(_id);

  @override
  bool allows(String deviceId) => _registry.host.allows(deviceId, _id);

  @override
  void emit(String name, Object? payload) =>
      _registry.host.emit(PluginEvent(_id, name, payload));

  @override
  Future<HandshakeResult> dialTransfer(String deviceId) =>
      _registry.host.dialTransfer(deviceId);

  @override
  Map<String, dynamic> settings() => _registry.host.loadPluginSettings(_id);

  @override
  Future<void> saveSettings(Map<String, dynamic> value) =>
      _registry.host.savePluginSettings(_id, value);

  @override
  bool get enabled => _registry.isEnabled(_id);

  @override
  void logf(String message) {
    // ignore: avoid_print
    print('[$_id] $message');
  }
}

File: core/plugin/registry.go
package plugin

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"

	"wedrop/core/protocol"
)

// Host is what the registry needs from whatever embeds core (the desktop
// Wails service today; a future Linux/macOS host later) to actually reach
// peers, persist settings, and surface events. The registry is otherwise
// self-contained and knows nothing about Wails, sockets, or storage
// formats.
type Host interface {
	Send(deviceID string, v interface{}) error
	Broadcast(capability string, v interface{})
	ConnectedPeers(capability string) []PeerRef
	Emit(event Event)
	LoadPluginSettings(id ID) []byte
	SavePluginSettings(id ID, data []byte) error
}

type registeredPlugin struct {
	plugin  Plugin
	enabled bool
}

// Registry holds every compiled-in plugin, routes inbound messages to the
// plugin that claimed each message type, and lets a plugin be enabled or
// disabled at runtime without touching any other plugin.
//
// This is the "loadable/unloadable" mechanism: plugins are linked into the
// binary at compile time (RegisterAll-style call sites own the flat list),
// but the registry's enable/disable is a genuine runtime toggle — a
// disabled plugin's messages are silently dropped and its background work
// is stopped, without restarting the process or any other plugin.
type Registry struct {
	host Host

	mu      sync.RWMutex
	plugins map[ID]*registeredPlugin
	byMsg   map[protocol.MessageType]ID
}

// NewRegistry creates an empty registry bound to host.
func NewRegistry(host Host) *Registry {
	return &Registry{
		host:    host,
		plugins: make(map[ID]*registeredPlugin),
		byMsg:   make(map[protocol.MessageType]ID),
	}
}

// Register adds a plugin, enabled by default unless enabledByDefault is
// false. It is an error for two plugins to claim the same message type.
func (r *Registry) Register(p Plugin, enabledByDefault bool) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	id := p.ID()
	if _, exists := r.plugins[id]; exists {
		return fmt.Errorf("plugin %q already registered", id)
	}
	for _, mt := range p.MessageTypes() {
		if owner, taken := r.byMsg[mt]; taken {
			return fmt.Errorf("message type %q already claimed by plugin %q", mt, owner)
		}
	}

	if err := p.Init(&pluginAPI{id: id, registry: r}); err != nil {
		return fmt.Errorf("init plugin %q: %w", id, err)
	}

	for _, mt := range p.MessageTypes() {
		r.byMsg[mt] = id
	}
	r.plugins[id] = &registeredPlugin{plugin: p, enabled: enabledByDefault}
	return nil
}

// Plugin returns a registered plugin by ID, if any.
func (r *Registry) Plugin(id ID) (Plugin, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	rp, ok := r.plugins[id]
	if !ok {
		return nil, false
	}
	return rp.plugin, true
}

// IDs lists every registered plugin ID.
func (r *Registry) IDs() []ID {
	r.mu.RLock()
	defer r.mu.RUnlock()
	ids := make([]ID, 0, len(r.plugins))
	for id := range r.plugins {
		ids = append(ids, id)
	}
	return ids
}

// Enabled reports whether a plugin is currently enabled. An unknown ID
// reports false.
func (r *Registry) Enabled(id ID) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	rp, ok := r.plugins[id]
	return ok && rp.enabled
}

// SetEnabled toggles a plugin at runtime. Disabling stops its background
// work and starts silently dropping its messages; re-enabling restarts it.
func (r *Registry) SetEnabled(ctx context.Context, id ID, enabled bool) error {
	r.mu.Lock()
	rp, ok := r.plugins[id]
	if !ok {
		r.mu.Unlock()
		return fmt.Errorf("unknown plugin %q", id)
	}
	already := rp.enabled
	rp.enabled = enabled
	r.mu.Unlock()

	if already == enabled {
		return nil
	}
	if enabled {
		return rp.plugin.Start(ctx)
	}
	rp.plugin.Stop()
	return nil
}

// StartAll starts every enabled plugin's background work.
func (r *Registry) StartAll(ctx context.Context) error {
	r.mu.RLock()
	toStart := make([]Plugin, 0, len(r.plugins))
	for _, rp := range r.plugins {
		if rp.enabled {
			toStart = append(toStart, rp.plugin)
		}
	}
	r.mu.RUnlock()

	for _, p := range toStart {
		if err := p.Start(ctx); err != nil {
			return fmt.Errorf("start plugin %q: %w", p.ID(), err)
		}
	}
	return nil
}

// StopAll stops every plugin's background work, regardless of enabled state.
func (r *Registry) StopAll() {
	r.mu.RLock()
	all := make([]Plugin, 0, len(r.plugins))
	for _, rp := range r.plugins {
		all = append(all, rp.plugin)
	}
	r.mu.RUnlock()

	for _, p := range all {
		p.Stop()
	}
}

// --------------------------------------------- transport.SessionHandler

// OnMessage implements transport.SessionHandler (structurally — core/plugin
// does not import core/transport to avoid a dependency cycle; see
// desktop/service.go for the adapter that wires this in). It looks up which
// plugin claimed msgType and, if that plugin is enabled, hands it the raw
// payload; an unclaimed or disabled plugin's message is dropped silently.
func (r *Registry) OnMessage(peer PeerRef, msgType protocol.MessageType, raw []byte) {
	r.mu.RLock()
	id, claimed := r.byMsg[msgType]
	var rp *registeredPlugin
	if claimed {
		rp = r.plugins[id]
	}
	r.mu.RUnlock()

	if !claimed || rp == nil || !rp.enabled {
		return
	}
	_ = rp.plugin.HandleMessage(peer, msgType, raw)
}

// OnPeerConnected/OnPeerDisconnected fan out session lifecycle changes to
// every enabled plugin.
func (r *Registry) OnPeerConnected(peer PeerRef) {
	for _, p := range r.enabledPlugins() {
		p.OnPeerConnected(peer)
	}
}

func (r *Registry) OnPeerDisconnected(deviceID string) {
	for _, p := range r.enabledPlugins() {
		p.OnPeerDisconnected(deviceID)
	}
}

func (r *Registry) enabledPlugins() []Plugin {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]Plugin, 0, len(r.plugins))
	for _, rp := range r.plugins {
		if rp.enabled {
			out = append(out, rp.plugin)
		}
	}
	return out
}

// HandleTransferOffer routes an inbound transfer connection to whichever
// registered plugin implements TransferPlugin (today, only files). If no
// such plugin is registered or it is disabled, the caller should close conn.
func (r *Registry) HandleTransferOffer(conn TransferConn, peer protocol.DeviceInfo, offer protocol.TransferOffer) {
	r.mu.RLock()
	var target *registeredPlugin
	for _, rp := range r.plugins {
		if _, ok := rp.plugin.(TransferPlugin); ok {
			target = rp
			break
		}
	}
	r.mu.RUnlock()

	if target == nil || !target.enabled {
		conn.Close()
		return
	}
	target.plugin.(TransferPlugin).HandleTransferOffer(conn, peer, offer)
}

// --------------------------------------------------------- per-plugin API

// pluginAPI is the API handed to exactly one plugin; every method is scoped
// to that plugin's ID acting as its capability string.
type pluginAPI struct {
	id       ID
	registry *Registry
}

func (a *pluginAPI) Send(deviceID string, v interface{}) error {
	return a.registry.host.Send(deviceID, v)
}

func (a *pluginAPI) Broadcast(v interface{}) {
	a.registry.host.Broadcast(string(a.id), v)
}

func (a *pluginAPI) ConnectedPeers() []PeerRef {
	return a.registry.host.ConnectedPeers(string(a.id))
}

func (a *pluginAPI) Emit(name string, payload interface{}) {
	a.registry.host.Emit(Event{Plugin: a.id, Name: name, Payload: payload})
}

func (a *pluginAPI) Settings() []byte {
	return a.registry.host.LoadPluginSettings(a.id)
}

func (a *pluginAPI) SaveSettings(v interface{}) error {
	data, err := json.Marshal(v)
	if err != nil {
		return err
	}
	return a.registry.host.SavePluginSettings(a.id, data)
}

func (a *pluginAPI) Enabled() bool {
	return a.registry.Enabled(a.id)
}

func (a *pluginAPI) Logf(format string, args ...interface{}) {
	fmt.Printf("[%s] "+format+"\n", append([]interface{}{a.id}, args...)...)
}
